File: Anime Game Remap (for all users)/api/src/py/FixRaidenBoss2/model/iniresources/RemapIniResource.py
##### Credits

# ===== Anime Game Remap (AG Remap) =====
# Authors: Albert Gold#2696, NK#1321
#
# if you used it to remap your mods pls give credit for "Albert Gold#2696" and "Nhok0169"
# Special Thanks:
#   nguen#2011 (for support)
#   SilentNightSound#7430 (for internal knowdege so wrote the blendCorrection code)
#   HazrateGolabi#1364 (for being awesome, and improving the code)

##### EndCredits


##### ExtImports
import os
import shutil
from typing import Optional, Callable, Any, TYPE_CHECKING, Any, List
##### EndExtImports

##### LocalImports
from .IniResource import IniResource, IniFixResource, IniGroupedResource
from ..stats.CachedFileStats import CachedFileStats
from ..stats.RemapStats import RemapStats
from ...tools.files.FileDownload import FileDownload
from ..files.BlendFile import BlendFile
from ..buffers.BufElementType import BufElementType
from ..VGRemap import VGRemap


if (TYPE_CHECKING):
    from ..Mod import Mod
##### EndLocalImports


##### Script
class RemapIniResourceMixin():
    def srcEncounteredError(self, stats: RemapStats) -> bool:
        """
        Determines whether the resources have previously encountered an error

        Parameters
        ----------
        stats: :class:`RemapStats`
            The stats tracked by the remap process

        Returns
        -------
        :class:`bool`
            Whether the resource has encountered an error
        """

        pass

    def srcIsFixed(self, stats: RemapStats) -> bool:
        """
        Determines whether the resources were already fixed

        Parameters
        ----------
        stats: :class:`RemapStats`
            The stats tracked by the remap process

        Returns
        -------
        :class:`bool`
            Whether the resource was already fixed
        """

        pass

    def fixEncounteredError(self, stats: RemapStats) -> bool:
        """
        Determines whether the fixed resources have previously encountered an error

        Parameters
        ----------
        stats: :class:`RemapStats`
            The stats tracked by the remap process

        Returns
        -------
        :class:`bool`
            Whether the fixed resource has encountered an error
        """

        pass

    def fixIsFixed(self, stats: RemapStats) -> bool:
        """
        Determines whether the fixed resources were already fixed

        Parameters
        ----------
        stats: :class:`RemapStats`
            The stats tracked by the remap process

        Returns
        -------
        :class:`bool`
            Whether the fixed resource was already fixed
        """

        pass

    def fixExists(self, stats: RemapStats) -> bool:
        """
        Determines whether the fixed resources already exist in the OS

        Parameters
        ----------
        stats: :class:`RemapStats`
            The stats tracked by the remap process

        Returns
        -------
        :class:`bool`
            Whether the fixed resources already exist
        """

        pass

    def hasRequired(self) -> bool:
        """
        Determines whether all the necessary resources have been collected

        Returns
        -------
        :class:`bool`
            whether all the required resources are gathered
        """

        pass

    def remapFix(self, mod: "Mod", stats: RemapStats, *args, **kwargs) -> bool:
        """
        Fixes the resource for the overall remap process

        Parameters
        ----------
        mod: :class:`Mod`
            The mod to fix from

        stats: :class:`RemapStats`
            The stats tracked by the remap process

        \*args
            Any extra arguments to supply

        \*\*kwargs
            Any extra keyword arguments to supply

        Returns
        -------
        :class:`bool`
            Whether the resource was fixed
        """

        pass


class RemapIniResource(IniResource, RemapIniResourceMixin):
    """
    This class inherits from :class:`IniResource`

    Base class for some resource in a .ini file that is used by the overall remap process at :class:`RemapService`

    Parameters
    ----------
    type: :class:`str`
        The name for the type of resource

    iniFolderPath: :class:`str`
        The path to the folder of the .ini file

    srcPath: :class:`str`
        The file path to the resource

    fixFunc: Optional[Callable[[:class:`RemapIniResource`, ...], Any]]
        Custom function for fixing the resource :raw-html:`<br />` :raw-html:`<br />`

        **Default**: ``None``
    """
    
    def hasRequired(self):
        return True
    
    def fixExists(self, stats: RemapStats):
        return self.srcIsFixed(stats)
    
    def remapFix(self, mod: "Mod", stats: RemapStats, *args, **kwargs) -> bool:
        return self.fix(*args, **kwargs)
    

class RemapIniFixResource(IniFixResource, RemapIniResourceMixin):
    """
    This class inherits from :class:`IniFixResource`

    Base class for some resource to fix in a .ini file that is used by the overall remap process at :class:`RemapService`

    Parameters
    ----------
    type: :class:`str`
        The name for the type of resource

    iniFolderPath: :class:`str`
        The path to the folder of the .ini file

    srcPath: :class:`str`
        The file path to the resource

    fixedPath: :class:`str`
        The file path to the fixed resource

    fixFunc: Optional[Callable[[:class:`RemapIniFixResource`, :class:`Mod`, :class:`RemapStats`, ...], Any]]
        Custom function for fixing the resource :raw-html:`<br />` :raw-html:`<br />`

        The function takes in:

        #. The resource to fix
        #. The mod where the resource comes from
        #. The file stats to track :raw-html:`<br />` :raw-html:`<br />`

        **Default**: ``None``
    """
    
    def fixExists(self):
        return os.path.isfile(self.fixedPath)
    
    def hasRequired(self):
        return True
    
    def remapFix(self, mod: "Mod", stats: RemapStats, *args, **kwargs) -> bool:
        return self.fix(*args, **kwargs)
    

class RemapIniGroupedResource(IniGroupedResource, RemapIniResourceMixin):
    """
    This class inherits from :class:`IniGroupedResource`

    Base class for a group of resources to fix in a .ini file that is used by the overall remap process at :class:`RemapService`

    Parameters
    ----------
    name: :class:`str`
        The name of the group of resources

    resources: Optional[Dict[:class:`str`, Union[:class:`IniResource`, :class:`str`]]]
        The group of resources. 

        * The keys are the type of the resource
        * If :attr:`isBuilt` is set to ``True`` then all the values are expected to be of type :class:`IniResource`
          Otherwise, all the values are expected to be a string value representing the source file location for the resource :raw-html:`<br />` :raw-html:`<br />`

        **Default**: ``None``

    fixFunc: Optional[Callable[[:class:`IniGroupedResource`, ...], Any]]
        Custom function for fixing the resource :raw-html:`<br />` :raw-html:`<br />`

        **Default**: ``None``

    isBuilt: :class:`bool`
        Whether the grouped resource is ready to be fixed :raw-html:`<br />` :raw-html:`<br />`

        **Default**: ``True``
    """

    def remapFix(self, mod: "Mod", stats: RemapStats, *args, **kwargs) -> bool:
        return self.fix(*args, **kwargs)


class RemapIniDownload(RemapIniResource):
    """
    This class inherits from :class:`RemapIniResource`

    Class for some download resource in a .ini file that is used by the overall remap process at :class:`RemapService`

    Parameters
    ----------
    iniFolderPath: :class:`str`
        The path to the folder of the .ini file

    srcPath: :class:`str`
        The file path to the resource

    download: :class:`FileDownload`
        The downloader associated to the file

    type: :class:`str`
        The name for the type of resource :raw-html:`<br />` :raw-html:`<br />`

        **Default**: ``"download"``

    fixFunc: Optional[Callable[[:class:`RemapIniDownload`, :class:`Mod`, :class:`RemapStats`, ...], Any]]
        Custom function for fixing the resource :raw-html:`<br />` :raw-html:`<br />`

        The function takes in:

        #. The resource to fix
        #. The mod where the resource comes from
        #. The file stats to track :raw-html:`<br />` :raw-html:`<br />`

        **Default**: ``None``
    """

    def __init__(self, iniFolderPath: str, srcPath: str, download: FileDownload, type: str = "download", fixFunc = None):
        super().__init__(type, iniFolderPath, srcPath, fixFunc = fixFunc)
        self.download = download

    def srcEncounteredError(self, stats: RemapStats) -> bool:
        return self.srcPath in stats.download.skipped
    
    def srcIsFixed(self, stats: RemapStats) -> bool:
        return self.srcPath in stats.download.fixed
    
    def fixEncounteredError(self, stats: RemapStats):
        return self.srcEncounteredError(stats)
    
    def fixIsFixed(self, stats: RemapStats):
        return self.srcIsFixed(stats)
    
    def fixExists(self, stats: RemapStats):
        return self.srcIsFixed(stats)

    def _fix(self, downloadStats: CachedFileStats, *args, proxy: Optional[str] = None, 
            downloadHandler: Optional[Callable[[str], Any]] = None, 
            cacheHitHandler: Optional[Callable[[str], Any]] = None, **kwargs) -> bool:
        """
        Downloads the resource

        Parameters
        ----------
        downloadStats: :class:`CachedFileStats`
            The stats for the file download for the overall remap process at :class:`RemapService`

        proxy: Optional[:class:`str`]
            The link to the proxy server used for any internet network requests made :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``None``

        downloadHandler: Optional[Callable[[:class:`str`], Any]]
            The callback once the file has been downloaded. The function takes in the filepath to the download :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``None``

        cacheHitHandler: Optional[Callable[[:class:`str`], Any]]
            The callback once the file was copied from the cache instead of needing to be downloaded. The function takes in the filepath to the download :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``None``

        Returns
        -------
        :class:`bool`
            Whether the resource has been downloaded
        """

        downloadFolder = os.path.dirname(self.srcPath)
        rawDownloadFullPath, downloaded, downloadExisted =  self.download.get(downloadFolder, proxy = proxy)

        if (self.srcPath != rawDownloadFullPath):
            shutil.move(rawDownloadFullPath, self.srcPath)

        if (downloaded):
            downloadStats.addFixed(self.srcPath)
            if (downloadHandler is not None):
                downloadHandler(self.srcPath)
            
        else:
            downloadStats.addHit(self.srcPath)
            if (cacheHitHandler is not None):
                cacheHitHandler(self.srcPath)
    
    def remapFix(self, mod: "Mod", stats: RemapStats, *args, proxy: Optional[str] = None, **kwargs):
        """
        Fixes the resource for the overall remap process

        Parameters
        ----------
        mod: :class:`Mod`
            The mod to fix from

        stats: :class:`RemapStats`
            The stats tracked by the remap process

        proxy: Optional[:class:`str`]
            The link to the proxy server used for any internet network requests made :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``None`` 

        Returns
        -------
        :class:`bool`
            Whether the resource was fixed
        """

        return self.fix(stats.download, proxy = proxy, 
                        downloadHandler = lambda downloadPath: mod.print("log", f"Download successful at {downloadPath}"),
                        cacheHitHandler = lambda downloadPath: mod.print("log", f"Copied previous download to {downloadPath}"))
##### EndScript

File: Anime Game Remap (for all users)/api/src/cpp/core/src/model/iniresources/RemapIniResource.cpp
// ##### Credits

// ===== Anime Game Remap (AG Remap) =====
// Authors: Albert Gold#2696, NK#1321
//
// if you used it to remap your mods pls give credit for "Albert Gold#2696" and "Nhok0169"
// Special Thanks:
//   nguen#2011 (for support)
//   SilentNightSound#7430 (for internal knowdege so wrote the blendCorrection code)
//   HazrateGolabi#1364 (for being awesome, and improving the code)

// ##### EndCredits

#include "AGRemapCore/tools/files/FileService.h"
#include "AGRemapCore/model/iniresources/RemapIniResource.h"

#include <filesystem>
#include <system_error>
#include <utility>


namespace AGRemapCore {

    namespace {
        // Mirrors Python's shutil.move -- a plain rename first (cheap, the common case), falling
        // back to copy-then-remove if that fails (eg. the source/destination are on different
        // filesystems/drives, which std::filesystem::rename doesn't handle the way shutil.move does).
        void moveFile(const std::string& from, const std::string& to) {
            std::error_code renameError;
            std::filesystem::rename(FileService::strToPath(from), to, renameError);
            if (!renameError) {
                return;
            }

            std::filesystem::copy_file(FileService::strToPath(from), to, std::filesystem::copy_options::overwrite_existing);
            std::filesystem::remove(FileService::strToPath(from));
        }
    }

    bool RemapIniResourceMixin::srcEncounteredError(const RemapStats& stats) const {
        (void)stats;
        return false;
    }

    bool RemapIniResourceMixin::srcIsFixed(const RemapStats& stats) const {
        (void)stats;
        return false;
    }

    bool RemapIniResourceMixin::fixEncounteredError(const RemapStats& stats) const {
        (void)stats;
        return false;
    }

    bool RemapIniResourceMixin::fixIsFixed(const RemapStats& stats) const {
        (void)stats;
        return false;
    }

    bool RemapIniResourceMixin::fixExists(const RemapStats& stats) const {
        (void)stats;
        return false;
    }

    bool RemapIniResourceMixin::hasRequired() const {
        return false;
    }

    bool RemapIniResource::hasRequired() const {
        return true;
    }

    bool RemapIniResource::fixExists(const RemapStats& stats) const {
        return srcIsFixed(stats);
    }

    bool RemapIniFixResource::hasRequired() const {
        return true;
    }

    bool RemapIniFixResource::fixExists(const RemapStats& stats) const {
        (void)stats;
        return std::filesystem::exists(FileService::strToPath(fixedPath));
    }

    RemapIniDownload::RemapIniDownload(const std::string& iniFolderPath, const std::string& srcPath, std::unique_ptr<FileDownload> download,
                                        std::string type, std::function<bool(RemapIniDownload&, CachedFileStats&)> fixFunc):
        RemapIniResource(std::move(type), iniFolderPath, srcPath), download(std::move(download)), fixFunc(std::move(fixFunc)) {}

    bool RemapIniDownload::srcEncounteredError(const RemapStats& stats) const {
        return stats.download.skipped.contains(srcPath);
    }

    bool RemapIniDownload::srcIsFixed(const RemapStats& stats) const {
        return stats.download.fixed.contains(srcPath);
    }

    bool RemapIniDownload::fixEncounteredError(const RemapStats& stats) const {
        return srcEncounteredError(stats);
    }

    bool RemapIniDownload::fixIsFixed(const RemapStats& stats) const {
        return srcIsFixed(stats);
    }

    bool RemapIniDownload::fixExists(const RemapStats& stats) const {
        return srcIsFixed(stats);
    }

    bool RemapIniDownload::_fix(CachedFileStats& downloadStats, std::optional<std::string> proxy) {
        std::string downloadFolder = FileService::pathToStr(FileService::strToPath(srcPath).parent_path());
        auto [rawDownloadFullPath, downloaded, downloadExisted] = download->get(downloadFolder, proxy, downloadCache);
        (void)downloadExisted;

        if (srcPath != rawDownloadFullPath) {
            moveFile(rawDownloadFullPath, srcPath);

            // The run's cache was told where get() PUT the file, and it is no longer there.
            // Correct it, or every later use of this URL pays a failed copy and re-downloads --
            // a cache that costs something and returns nothing.
            if (downloaded && downloadCache != nullptr && download != nullptr) {
                downloadCache->remember(download->url, srcPath);
            }
        }

        if (downloaded) {
            downloadStats.addFixed(srcPath);
        } else {
            downloadStats.addHit(srcPath);
        }

        return downloaded;
    }

    bool RemapIniDownload::willDownload() const {
        if (download == nullptr || fixFunc) {
            return true;
        }

        return !download->cachedPath(downloadCache).has_value();
    }

    bool RemapIniDownload::fix(CachedFileStats& downloadStats, std::optional<std::string> proxy) {
        // BEFORE the work, for the reason RemapBlendResource::fix gives and then some: a download
        // is the slowest thing this program does and much the likeliest to fail, so this is the
        // line that explains a long pause and the one still on screen when the request throws
        // ("Could not resolve hostname" being the usual one).
        //
        // A cache hit gets its own wording rather than the same one: copying a file this run has
        // already fetched is instant and cannot fail the way a request can, so calling it
        // "Downloading" would make a 36-.ini mod look like 36 trips to github when 35 of them
        // never leave the disk. No trailing "..." on that one for the same reason -- there is
        // nothing to wait for.
        const std::string name = FileService::pathToStr(FileService::strToPath(srcPath).filename());
        const bool expectingDownload = willDownload();

        if (logger != nullptr) {
            logger->log(expectingDownload ? "Downloading " + name + "..."
                                          : "Copying download " + name);
        }

        const bool downloaded = fixFunc ? fixFunc(*this, downloadStats) : _fix(downloadStats, proxy);

        // The prediction was that a copy would do, and it did not -- FileDownload::get found the
        // remembered file gone and went to the network after all. Rare, but saying so is the
        // difference between a log that describes the run and one that describes an intention.
        if (!expectingDownload && downloaded && logger != nullptr) {
            logger->log("Downloading " + name + "...");
        }

        return downloaded;
    }

    bool RemapIniDownload::remapFix(RemapStats& stats, std::optional<std::string> proxy,
                                     const std::function<void(const std::string&)>& downloadHandler,
                                     const std::function<void(const std::string&)>& cacheHitHandler) {
        bool downloaded = fix(stats.download, proxy);

        if (downloaded) {
            if (downloadHandler) {
                downloadHandler(srcPath);
            }
        } else {
            if (cacheHitHandler) {
                cacheHitHandler(srcPath);
            }
        }

        return downloaded;
    }
}
